sw/device/lib/runtime/epmp.h

```diff
@@ -0,0 +1,458 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef OPENTITAN_SW_DEVICE_LIB_RUNTIME_EPMP_H_
+#define OPENTITAN_SW_DEVICE_LIB_RUNTIME_EPMP_H_
+
+#include <stdint.h>
+
+/**
+ * RISC-V Enhanced Physical Memory Protection (EPMP).
+ *
+ * Specifications:
+ *  - PMP Enhancements for memory access and execution prevention on Machine
+ *    mode
+ *    https://docs.google.com/document/d/1Mh_aiHYxemL0umN3GTTw8vsbmzHZ_nxZXgjgOUzbvc8
+ *  - RISC-V Privileged Specfication (v. 20190608)
+ *    https://github.com/riscv/riscv-isa-manual/releases/download/Ratified-IMFDQC-and-Priv-v1.11/riscv-privileged-20190608.pdf
+ *  - Ibex PMP
+ *    https://ibex-core.readthedocs.io/en/latest/03_reference/pmp.html
+ *
+ * Assumptions (should be initialized in assembly but can be verified using
+ * `epmp_check`):
+ *   * Machine Mode Whitelist Policy is enabled (mseccfg.MMWP = 1)
+ *   * Machine Mode Lockdown is disabled (mseccfg.MML = 0)
+ */
+
+/**
+ * EPMP constants.
+ *
+ * TODO: there will probably be more of these.
+ */
+typedef enum epmp_constants {
+  kEpmpGranularity = 0,
+  kEpmpNumRegions = 16,
+} epmp_constants_t;
+
+/**
+ * Toggle for Rule Locking Bypass (mseccfg.RLB).
+ */
+typedef enum epmp_rlb_toggle {
+  kEpmpRlbToggleDisabled = 0, /**< mseccfg.RLB = 0 */
+  kEpmpRlbToggleEnabled,      /**< mseccfg.RLB = 1 */
+} epmp_rlb_toggle_t;
+
+/**
+ * EPMP state.
+ *
+ * A copy of a possible EPMP control register state.
+ */
+typedef struct epmp_state {
+  uint8_t pmpcfg[kEpmpNumRegions];
+  uintptr_t pmpaddr[kEpmpNumRegions];
+} epmp_state_t;
+
+/**
+ * EPMP entry permissions.
+ *
+ * Unlocked permissions can generally be modified when in M-mode. Locked
+ * permissions can only be modified in M-mode if Rule Locking Bypass
+ * (mseccfg.RLB) is set.
+ *
+ * When Machine Mode Lockdown is disabled (mseccfg.MML is unset) the
+ * combination R=0 W=1 is reserved. Note that this is the assumed state
+ * and so it is not possible to set these values.
+ *
+ * Note: these permissions have different meanings when Machine Mode
+ * Lockdown (mseccfg.MML) is set.
+ *
+ * TODO: MML mode could be supported by adding additional enums. That
+ * is why M-mode and S/U-mode permissions have been made explicit in
+ * enum names.
+ */
+typedef enum epmp_permissions {
```

Comment:
Hmm, should this be a bitfield, or perhaps some kind of struct? It feels a bit hard to read the name of the constants because they're so long. =/

Maybe it's just me but I feel like `kRead | kWrite | kLocked` is a bit easier to follow?

Reply:
I agree, I prefer shorter names as well. I think enums are fine if we can come up with some shorter names, but bitfields are also an option.

Reply:
In general (and this applies to some of the other comments on other enums as well) this API is using enums specifically so that they can be hardened later. My hope is that using hardened enum values we can avoid adding as many integrity checks to the parameters (e.g. checksum or redundant copies) that might make the API more difficult to use correctly.

> Maybe it's just me but I feel like kRead | kWrite | kLocked is a bit easier to follow?

Yes, it might be in some situations. Although there are also good reasons to use an enum:
1. some combinations are deliberately left out (R=0 W=1)
2. more descriptive names make it clearer what the permissions actually are rather than the raw bit values (this would admittedly matter more if we supported MML where the mapping is not 1:1)

We can probably make the names shorter. We can also reduce the number if we are happy to make some assumptions (e.g. we might assume that we are always in machine mode and therefore don't need unlocked rules in this code).